onnxruntime/core/providers/trt/trt_execution_provider.cc

```diff
@@ -0,0 +1,420 @@
+    trt_parser->parse(onnx_buf.data(), onnx_buf.size());
+    trt_builder->setMaxBatchSize(max_batch_size);
+    trt_builder->setMaxWorkspaceSize(max_workspace_size);
+    const auto& trt_engine = InferObject(trt_builder->buildCudaEngine(*trt_network.get()));
+    ORT_ENFORCE(trt_engine != nullptr);
+
+    // Build TensorRT context
+    const auto& trt_context = InferObject(trt_engine->createExecutionContext());
+    ORT_ENFORCE(trt_context != nullptr);
+
+    // Get input shape and binding index
+    input_shape.resize(trt_network->getNbInputs());
+    for (int i = 0, end = trt_network->getNbInputs(); i < end; ++i) {
+      const std::string& name = trt_network->getInput(i)->getName();
+      size_t bindingIndex = trt_engine->getBindingIndex(name.c_str());
+
+      nvinfer1::Dims4 dimensions = static_cast<nvinfer1::Dims4&&>(trt_engine->getBindingDimensions(static_cast<int>(bindingIndex)));
+      size_t dim_size = 1;
+      for (int j = 0, end = dimensions.nbDims; j < end; ++j) {
+        input_shape[i].push_back(dimensions.d[j]);
+        dim_size *= dimensions.d[j];
+      }
+
+      input_binding_indexes.push_back(bindingIndex);
+      input_dim_sizes.push_back(dim_size);
+    }
+
+    // Get output shape and binding index
+    output_shapes.resize(trt_network->getNbOutputs());
+    for (int i = 0, end = trt_network->getNbOutputs(); i < end; ++i) {
+      const std::string& name = trt_network->getOutput(i)->getName();
+      size_t bindingIndex = trt_engine->getBindingIndex(name.c_str());
+
+      nvinfer1::Dims4 dimensions = static_cast<nvinfer1::Dims4&&>(trt_engine->getBindingDimensions(static_cast<int>(bindingIndex)));
+      size_t dim_size = 1;
+      for (int j = 0, end = dimensions.nbDims; j < end; ++j) {
+        output_shapes[i].push_back(dimensions.d[j]);
+        dim_size *= dimensions.d[j];
+      }
+
+      output_binding_indexes.push_back(bindingIndex);
+      output_dim_sizes.push_back(dim_size);
+    }
+
+    int num_inputs = input_binding_indexes.size();
+    int num_outputs = output_binding_indexes.size();
+    ORT_ENFORCE(trt_engine->getNbBindings() == (num_inputs + num_outputs));
+
+    // Save engine, context and input/output info to map
+    parsers_[fused_node->Name()] = trt_parser;
+    engines_[fused_node->Name()] = trt_engine;
+    contexts_[fused_node->Name()] = trt_context;
+    input_info_[fused_node->Name()].push_back(input_indexes);
+    input_info_[fused_node->Name()].push_back(input_binding_indexes);
+    input_info_[fused_node->Name()].push_back(input_dim_sizes);
+    output_info_[fused_node->Name()].push_back(output_binding_indexes);
+    output_info_[fused_node->Name()].push_back(output_dim_sizes);
+    output_shapes_[fused_node->Name()] = output_shapes;
+
+    // Create function state
+    NodeComputeInfo compute_info;
+    compute_info.create_state_func = [=](ComputeContext* context, FunctionState* state) {
+      auto* p = new TRTFuncState();
```

Comment:
Use make_unique to create this ptr and then use ```*state = p.release()```. This is exception safe and will ensure the ptr is deleted in case an exception is thrown between this line and the state assignment.